Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,333 @@
+        0, G_MAXUINT, 0,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    /**
+     * WebKitImage:pixels:
+     *
+     * The image pixels
+     */
+    sObjProperties[PROP_PIXELS] =
+    g_param_spec_boxed(
+        "pixels",
+        nullptr, nullptr,
+        G_TYPE_BYTES,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    g_object_class_install_properties(objectClass, N_PROPERTIES, sObjProperties.data());
+}
+
+/**
+ * webkit_image_new:
+ * @width: the image width
+ * @height: the image height
+ * @stride: the image stride
+ * @pixels: the image pixels
+ *
+ * Create a new #WebKitImage for the given parameters
+ *
+ * Returns: (transfer full): a #WebKitImage
+ */
+WebKitImage* webkit_image_new(int width, int height, guint stride, GBytes* pixels)
+{
+    g_return_val_if_fail(pixels, nullptr);
+
+    return WEBKIT_IMAGE(g_object_new(WEBKIT_TYPE_IMAGE,
+        "width", width,
+        "height", height,
+        "stride", stride,
+        "pixels", pixels,
+        nullptr));
+}
+
+/**
+ * webkit_image_get_width:
+ * @image: a #WebKitImage
+ *
+ * Get the @image width
+ *
+ * Returns: the image width
+ */
+int webkit_image_get_width(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), 0);
+    return image->priv->width;
+}
+
+/**
+ * webkit_image_get_height:
+ * @image: a #WebKitImage
+ *
+ * Get the @image height
+ *
+ * Returns: the image height
+ */
+int webkit_image_get_height(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), 0);
+    return image->priv->height;
+}
+
+/**
+ * webkit_image_get_stride:
+ * @image: a #WebKitImage
+ *
+ * Get the @image stride
+ *
+ * Returns: the image stride
+ */
+guint webkit_image_get_stride(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), 0);
+    return image->priv->stride;
+}
+
+/**
+ * webkit_image_get_pixels:
+ * @image: a #WebKitImage
+ *
+ * Get the @image pixel
+ *
+ * Returns: (transfer none): a #GBytes
+ */
+GBytes* webkit_image_get_pixels(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), nullptr);
+    return image->priv->pixels.get();
+}
+
+static void webkit_image_gicon_interface_init(GIconIface* iface)
+{
+    iface->hash = webkit_image_hash;
+    iface->equal = webkit_image_equal;
+}
+
+static void webkit_image_gloadable_icon_interface_init(GLoadableIconIface* iface)
+{
+    iface->load = webkit_image_load;
+    iface->load_async = webkit_image_load_async;
+    iface->load_finish = webkit_image_load_finish;
+}
+
+guint webkit_image_hash(GIcon* icon)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon), 0);
+
+    auto* image = WEBKIT_IMAGE(icon);
+    guint hash = g_bytes_hash(image->priv->pixels.get());
+    hash = hash * 31 + static_cast<guint>(image->priv->width);
+    hash = hash * 31 + static_cast<guint>(image->priv->height);
+    hash = hash * 31 + static_cast<guint>(image->priv->stride);
+    return hash;
```

Comment:
Probably better to use `WTF::Hasher` and/or `WTF::computeHash()`, which is type-aware?

```suggestion
    return computeHash(g_bytes_hash(webkit_image_as_rgba_bytes()), image->priv->width, image->priv->height);
```

Also I think that the stride should not matter, two images will display the same as long as their width, height, and pixel contents match&mdash;regardless of how the pixel data is distributed in memory.

Reply:
I didn't know about `computeHash` but it feels better, and agree about removing the stride